prov/lnx/src/lnx_ops.c

```diff
@@ -0,0 +1,1036 @@
+ * providers.
+ *
+ * Therefore when we check the unexpected queue, we need to check
+ * if we received any messages from any of the peer's addresses. If we
+ * find one, then we kick the core provider associated with that
+ * address to receive the message.
+ *
+ * If nothing is found on the unexpected messages, then add a receive
+ * request on the SRQ; happens in the lnx_process_recv()
+ */
+static int lnx_process_recv(struct lnx_ep *lep, struct iovec *iov, void **desc,
+			fi_addr_t addr, size_t count, struct lnx_peer *lp, uint64_t tag,
+			uint64_t ignore, void *context, uint64_t flags,
+			bool tagged)
+{
+	struct lnx_peer_srq *lnx_srq = &lep->le_srq;
+	struct local_prov_ep *cep;
+	struct lnx_rx_entry *rx_entry;
+	struct lnx_match_attr match_attr;
+	int rc = 0;
+
+	match_attr.lm_addr = addr;
+	match_attr.lm_ignore = ignore;
+	match_attr.lm_tag = tag;
+	match_attr.lm_cep = NULL;
+	match_attr.lm_peer = lp;
+
+	/* if support is turned off, don't go down the SRQ path */
+	if (!lep->le_domain->ld_srx_supported)
+		return -FI_ENOSYS;
+
+	rx_entry = lnx_remove_first_match(&lnx_srq->lps_trecv.lqp_unexq,
+					  &match_attr);
+	if (!rx_entry) {
+		FI_DBG(&lnx_prov, FI_LOG_CORE,
+		       "addr=%lx tag=%lx ignore=%lx buf=%p len=%lx not found\n",
+		       addr, tag, ignore, iov->iov_base, iov->iov_len);
+
+		goto nomatch;
+	}
+
+	FI_DBG(&lnx_prov, FI_LOG_CORE,
+	       "addr=%lx tag=%lx ignore=%lx buf=%p len=%lx found\n",
+	       addr, tag, ignore, iov->iov_base, iov->iov_len);
+
+	cep = rx_entry->rx_cep;
+
+	/* match is found in the unexpected queue. call into the core
+	 * provider to complete this message
+	 */
+	lnx_init_rx_entry(rx_entry, iov, desc, count, addr, tag, ignore,
+			  context, lnx_ep_rx_flags(lep));
+	rx_entry->rx_entry.msg_size = MIN(ofi_total_iov_len(iov, count),
+				      rx_entry->rx_entry.msg_size);
+	if (tagged)
+		rc = cep->lpe_srx.peer_ops->start_tag(&rx_entry->rx_entry);
+	else
+		rc = cep->lpe_srx.peer_ops->start_msg(&rx_entry->rx_entry);
+
+	if (rc == -FI_EINPROGRESS) {
+		/* this is telling me that more messages can match the same
+		 * rx_entry. So keep it on the queue
+		 */
+		FI_DBG(&lnx_prov, FI_LOG_CORE,
+		       "addr = %lx tag = %lx ignore = %lx start_tag() in progress\n",
+		       addr, tag, ignore);
+
+		goto insert_recvq;
+	} else if (rc) {
+		FI_WARN(&lnx_prov, FI_LOG_CORE, "start tag failed with %d\n", rc);
+	}
+
+	FI_DBG(&lnx_prov, FI_LOG_CORE,
+	       "addr = %lx tag = %lx ignore = %lx start_tag() success\n",
+	       addr, tag, ignore);
+
+	return 0;
+
+nomatch:
+	/* nothing on the unexpected queue, then allocate one and put it on
+	 * the receive queue
+	 */
+	rx_entry = get_rx_entry(NULL, iov, desc, count, addr, tag, ignore,
+				context, lnx_ep_rx_flags(lep));
+	rx_entry->rx_entry.msg_size = ofi_total_iov_len(iov, count);
+	if (!rx_entry) {
+		rc = -FI_ENOMEM;
+		goto out;
+	}
+	rx_entry->rx_peer = lp;
+
+insert_recvq:
+	lnx_insert_rx_entry(&lnx_srq->lps_trecv.lqp_recvq, rx_entry);
```

Comment:
I'm concerned the srx implementation doesn't handle FI_ADDR_UNSPEC properly. If an incoming message calls get_msg() with FI_ADDR_UNSPEC. We can't match it to any posted receive. We can only match it to a receive posted with FI_ADDR_UNSPEC or we have to wait until the address for that message is updated. I don't see any separation of those concepts. Need to read through the srx implementation again to try to follow.